### PR TITLE
(FACT-1238) Respect fully-qualified hostnames if specified.

### PR DESCRIPTION
**Pull request by peterhuene**
Facter 2.x would respect any hostname that was already fully-qualified.  A
hostname of 'foo.bar.com' would be split into a hostname of 'foo' and a domain
of 'bar.com'. No further attempt to resolve the FQDN would take place.

Facter 3 always used fallback logic of attempting to resolve the hostname to a
FQDN and looking through `resolv.conf` for domains.  This should only occur if
the system hostname is not already fully-qualified.

This is to say, if the users have configured their machine hostnames to be
fully-qualified, Facter should respect that.

### CODE REVIEW

File: lib/src/facts/posix/networking_resolver.cc [hunk 1]
```diff
@@ -106,7 +110,7 @@ namespace facter { namespace facts { namespace posix {
                     result.domain.assign(parts[1].begin(), parts[1].end());
                     return false;
                 }
-                if (parts[0] == boost::as_literal("search")) {
+                if (search.empty() && parts[0] == boost::as_literal("search")) {
```
Comment (MikaelSmith) [inline lib/src/facts/posix/networking_resolver.cc]: Did this fix a different problem?

File: lib/src/facts/posix/networking_resolver.cc [hunk 2]
```diff
@@ -67,15 +67,19 @@ namespace facter { namespace facts { namespace posix {
         if (gethostname(name.data(), name.size()) != 0) {
             LOG_WARNING("gethostname failed: %1% (%2%): hostname is unavailable.", strerror(errno), errno);
         } else {
-            // Use everything up to the first period
-            result.hostname = name.data();
-            auto pos = result.hostname.find('.');
-            if (pos != string::npos) {
-                result.hostname = result.hostname.substr(0, pos);
+            // Check for fully-qualified hostname
```
Comment (MikaelSmith) [inline lib/src/facts/posix/networking_resolver.cc]: Maps to approximately https://github.com/puppetlabs/facter/blob/2.x/lib/facter/domain.rb#L45.

File: lib/src/facts/posix/networking_resolver.cc (hunk 1 above)
Comment (peterhuene) [inline lib/src/facts/posix/networking_resolver.cc]: Noticed a discrepancy between the comment and the code: it was respecting the last `search` directive and not the first.  Can split into a separate commit if desired.

File: lib/src/facts/posix/networking_resolver.cc [hunk 3]
```diff
@@ -67,15 +67,19 @@ namespace facter { namespace facts { namespace posix {
...
+            auto it = find(name.begin(), name.end(), '.');
+            if (it != name.end()) {
+                LOG_DEBUG("using the FQDN returned by gethostname: %1%.", name.data());
+                result.hostname.assign(name.begin(), it);
+                result.domain.assign(++it, name.end());
```
Comment (peterhuene) [inline lib/src/facts/posix/networking_resolver.cc]: Actually, this is going to insert extra nulls into `domain` (prints ok, but the string shouldn't have nulls in it). Let me push up a fix.

File: lib/src/facts/posix/networking_resolver.cc (hunk 1 above)
Comment (peterhuene) [inline lib/src/facts/posix/networking_resolver.cc]: Since I have to push up an amended commit anyway, let me pull this out into its own commit.

**Comment by peterhuene** [comment]
@MikaelSmith Pushed up changes.
